### DISCUSSION TITLE
Publish message doesnt work when calling connect and publish in diff function

### DISCUSSION THREAD

**Discussion by adiedjx** [nats-io/nats.py#291] [Q&A]
```
import asyncio
import nats


async def working():
    nc = await nats.connect("nats://localhost:4222")
    st = bytes('str', 'utf-8')
    await nc.publish("subj", st)
    # await nc.drain()


def init():
    nc = asyncio.run(nats.connect("nats://localhost:4222"))
    return nc


def publish(nc):
    st = bytes('str', 'utf-8')
    asyncio.run(nc.publish("subj", st))


def not_working():
    nc = init()
    publish(nc)


if __name__ == '__main__':
    asyncio.run(working())
    not_working()

```

As per the code block, when i am initialising the nats and publishing in diff function the message isnt received on consumer, however it works when i init and publish in one function 

EDIT:Actually i was trying to initialise nats connection in the __init__ function and reuse the created connection while publishing, however i cant use await in __init__, from the various examples i saw, its like we have to initialize the connection each time before publishing a new message, i am not sure if that's the right approach

**Comment by KMilhan** [comment]
AFAIK, [`asyncio.run`](https://docs.python.org/3/library/asyncio-task.html#asyncio.run) is for the top-level function. i.e., `main`.  NATS internally creates clients per event loop, that could also be why it doesn't recognize initialized clients.

Try
```python
async def init() -> nats.NATS:
    nc = await nats.connect("nats://localhost:4222")
    return nc


async def publish(nc: nats.NATS):
    st = bytes("str", "utf-8")
    await nc.publish("subj", st)


async def not_working():
    nc = await init()
    await publish(nc)
    await nc.drain()


if __name__ == "__main__":
    asyncio.run(not_working())
```

**Comment by wallyqs** [comment]
something like the following would work:
```python
import asyncio
import nats

async def main():
    nc = await nats.connect("tls://demo.nats.io:4443")
    await nc.publish("test", b'hello')
    await nc.close() # so that buffer is flushed and sent

if __name__ == '__main__':
    asyncio.run(main())
```

**Comment by adiedjx** [comment]
Thanks, but actually i was trying to initialise nats connection in the __init__ function and reuse the created connection while publishing, however i cant use await in __init__, from the various examples i saw, its like we have to initialize the connection each time before publishing a new message, i am not sure if that's the right approach

How can i reuse the connection after initialising it once

**Comment by wallyqs** [comment]
Added an example of wrapping a class for the connection to reuse it later here: https://github.com/nats-io/nats.py/blob/main/examples/component.py
